### PR TITLE
Materials permissions

### PR DESCRIPTION
**Pull request by bandvov**
should be merged after materials-pagination

### CODE REVIEW

File: __tests__/model/model.mutation.test.js [hunk 1]
```diff
@@ -28,254 +34,264 @@ describe('Product queries', () => {
...
+              statusCode
+              message
+            }
           }
-        `,
-        variables: { id: "56ade69dd46eafc5968e5390" },
-      });
+        }
+      `,
+      variables: { id: '56ade69dd46eafc5968e5390' },
```
Comment (VolodymyrTrach) [inline __tests__/model/model.mutation.test.js:167]: Please move this id to const wrongId = '56ade69dd46eafc5968e5390' and use it here

File: __tests__/product/product.mutation.test.js [hunk 2]
```diff
@@ -12,7 +12,8 @@ const {
 
 require('dotenv').config();
 
-let productId, categoryId, subcategoryId, modelId, sameNameProductId;
+let productId; let categoryId; let subcategoryId; let modelId; let
+    sameNameProductId;
```
Comment (VolodymyrTrach) [inline __tests__/product/product.mutation.test.js:16]: Why did you change it? Please change it back

File: __tests__/product/product.query.test.js [hunk 3]
```diff
@@ -1,10 +1,16 @@
...
+  newModel,
+  badProductId,
+  newCategory,
+  getNewProduct,
+} = require('./product.variables');
 require('dotenv').config();
 
-let productId, categoryId, subcategoryId, modelId
+let productId; let categoryId; let subcategoryId; let
+    modelId;
```
Comment (VolodymyrTrach) [inline __tests__/product/product.query.test.js:13]: Is this ESlint or you? Please change it back

Comment (bandvov) [reply __tests__/product/product.query.test.js:13]: it is formatted by eslint

File: __tests__/product/product.mutation.test.js (hunk 2 above)
Comment (bandvov) [reply __tests__/product/product.mutation.test.js:16]: it is formatted by eslint

File: modules/comment/comment.service.js [hunk 4]
```diff
@@ -50,26 +50,26 @@ class CommentsService {
     const product = await Product.findById(id);
     const { userRates } = product;
     let { rateCount } = product;
-    const { rate } = userRates.find(rate => String(rate.user) === String(user._id)) || {};
+    const { rate } =      userRates.find(rate => String(rate.user) === String(user._id)) || {};
```
Comment (Alan18081) [inline modules/comment/comment.service.js:53]: I think it is better to have two different flows, where rate exists and where we need to create a rate object from scratch, it will be much more readable

File: __tests__/materials/material.mutation.test.js [hunk 5]
```diff
@@ -0,0 +1,440 @@
...
+            lang: 'uk',
+            value: 'проста назва кольору',
+          },
+          { __typename: 'Language', lang: 'en', value: 'simple color name' },
+        ],
+      },
+    ]);
+    expect(newMaterial.colors).toBeInstanceOf(Array);
+  });
+  test('#2 adding existing material should return error', async () => {
```
Comment (Alan18081) [inline __tests__/materials/material.mutation.test.js:144]: Please, write here something like `adding material with the existing name`, it ii much better reflects validation

File: __tests__/materials/material.mutation.test.js [hunk 6]
```diff
@@ -0,0 +1,440 @@
...
+
+let token;
+let materialId;
+const materialDoesNotExistId = '1f2ad410eb01783384e6111b';
+
+describe('test material mutations', () => {
+  beforeAll(async () => {
+    token = await adminLogin(user);
+  });
+  test('#1 should add material to database', async () => {
```
Comment (Alan18081) [inline __tests__/materials/material.mutation.test.js:25]: Please, create 3 `describe` statements for `Create` tests, `Update` tests, `Delete` tests to increase the readability of the tests

File: __tests__/materials/material.mutation.test.js [hunk 7]
```diff
@@ -0,0 +1,440 @@
...
+  materialToUpdate,
+  materialWithExistingName,
+} = require('./material.variables');
+const { adminLogin } = require('../helper-functions');
+
+let token;
+let materialId;
+const materialDoesNotExistId = '1f2ad410eb01783384e6111b';
+
+describe('test material mutations', () => {
```
Comment (Alan18081) [inline __tests__/materials/material.mutation.test.js:21]: Please, also add here tests for permissions, that simple users without admin permissions cannot add, update or delete materials and check that proper error messages are returned.

File: modules/comment/comment.service.js (hunk 4 above)
Comment (bandvov) [reply modules/comment/comment.service.js:53]: it is not my component

File: __tests__/materials/material.mutation.test.js (hunk 5 above)
Comment (bandvov) [reply __tests__/materials/material.mutation.test.js:144]: fixed

File: __tests__/materials/material.mutation.test.js (hunk 6 above)
Comment (bandvov) [reply __tests__/materials/material.mutation.test.js:25]: fixed

File: __tests__/materials/material.mutation.test.js (hunk 7 above)
Comment (bandvov) [reply __tests__/materials/material.mutation.test.js:21]: fixed

File: __tests__/model/model.mutation.test.js (hunk 1 above)
Comment (bandvov) [reply __tests__/model/model.mutation.test.js:167]: fixed

File: __tests__/materials/material.mutation.test.js [hunk 8]
```diff
@@ -18,11 +18,11 @@ let token;
 let materialId;
 const materialDoesNotExistId = '1f2ad410eb01783384e6111b';
 
-describe('test material mutations', () => {
-  beforeAll(async () => {
-    token = await adminLogin(user);
-  });
-  test('#1 should add material to database', async () => {
+beforeAll(async () => {
```
Comment (Alan18081) [inline __tests__/materials/material.mutation.test.js:21]: You still need root `describe` statement with some name like `Materials mutations tests`

File: __tests__/materials/material.mutation.test.js [hunk 9]
```diff
@@ -257,8 +258,8 @@ describe('test material mutations', () => {
 
     expect(updatedMaterial).toBeDefined();
     expect(updatedMaterial).toHaveProperty('name', [
-      { __typename: 'Language', lang: 'uk', value: 'Тест update' },
-      { __typename: 'Language', lang: 'en', value: 'Test update' },
+      { __typename: 'Language', lang: 'uk', value: 'Тест mu ta tion' },
```
Comment (Alan18081) [inline __tests__/materials/material.mutation.test.js:261]: Please, try to compare returned data with the data from the `materialToUpdate` object to avoid duplicated data and reduce the chance of mismatch errors if someone changes the data inside `materialToUpdate`

Comment (Alan18081) [reply __tests__/materials/material.mutation.test.js:261]: Please, do it for all fields in materials tests

Comment (bandvov) [reply __tests__/materials/material.mutation.test.js:261]: now it is impossible to do because we receive   __typenames in objects. and in  `materialToUpdate` there is no __typenames

File: __tests__/materials/material.mutation.test.js (hunk 8 above)
Comment (bandvov) [reply __tests__/materials/material.mutation.test.js:21]: fixed

File: __tests__/materials/material.mutation.test.js (hunk 9 above)
Comment (bandvov) [reply __tests__/materials/material.mutation.test.js:261]: fixed

Comment (bandvov) [reply __tests__/materials/material.mutation.test.js:261]: fixed
